### PR TITLE
Improve log message on plugin load fail/Decouple failure handling from task execution

### LINKED ISSUE #372: Improve log message when plugin(s) cannot be loaded
## The problem
Currently the message just says `java.lang.RuntimeException: 1 plugins could not be loaded`

## Detailed description
The current message is not useful because it does not tell us which message or why

## Acceptance criteria
* Include a list of names of plugins that failed
* Include at least the exception messages for each failed plugin

### PR DESCRIPTION
**Pull request by kvosper**
This PR introduces a common class called `FailureHandlingStrategy` that can be used to decouple failure-handling logic from the logic of processing items in a list.

Demonstrating this decoupling are the changes in `PluginSuppliers`, which also fixes #372.

### CODE REVIEW

File: components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                onEachFailure.accept(input, t);
+                failures.put(input, t);
+            }
+        });
+
+        if (!failures.isEmpty()) {
+            failuresPostProcessing.accept(failures);
+        }
+
+        return successes;
```
Comment (dvlato) [inline components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java]: do you think it would make sense to change the default behaviour to 'failure' instead of just returning the number of errors?

File: components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        public Builder<T, R> doImmediatelyOnEachFailure(BiConsumer<T, Exception> onEachFailure) {
+            this.onEachFailure = requireNonNull(onEachFailure);
+            return this;
+        }
+
+        /**
+         * An action to take after processing all list items, if there were failures.
+         * This will only execute if the individual failures were did not rethrow their exceptions.
+         *
+         * This action may be used to throw an exception, otherwise the {@link #process(List, Task)} method
```
Comment (dvlato) [inline components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java]: I would say that documenting what happens in 'external' code (the _otherwise_ part) should generally be avoided. You can just maybe rephrase the sentence.

File: components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    /**
+     * A task to execute. Similar to {@link java.util.function.Function}, except it is allowed to throw any type of
+     * {@link Exception}.
+     *
+     * @param <T> input type
+     * @param <R> output type
+     */
+    public interface Task<T, R> {
```
Comment (dvlato) [inline components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java]: If I were to read this the code for the first time, I would probably interpret 'Task' as something to do with concurrency... If you agree it could be misleading, can you think of a better name?

File: components/proxy/src/main/java/com/hotels/styx/proxy/plugin/PluginSuppliers.java [hunk 4]
```diff
@@ -44,11 +46,20 @@
     private final PluginFactoryLoader pluginFactoryLoader;
     private final Environment environment;
 
+    private final FailureHandlingTaskProcessor<Pair<String, SpiExtension>, NamedPlugin> processor =
```
Comment (dvlato) [inline components/proxy/src/main/java/com/hotels/styx/proxy/plugin/PluginSuppliers.java]: The name of the variable could probably be improved...

File: components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java (hunk 2 above)
Comment (kvosper) [reply components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java]: Is it really external if it is in the parent class that the builder creates?
